### ISSUE TITLE
Cannot run with aribtrary user ID

### ISSUE THREAD

**Issue by mirekphd** [rocker-org/rocker#295]
The image cannot be run with arbitrary user ID (unknown ex-ante, possibly random, as enforced by OpenShift default security policy). 

To reproduce set -u switch of docker run to anything other than 0:
```
docker run -p 8787:8787 -e USER=rstudio -e PASSWORD=rstudio -u 123456 rocker/rstudio
```

Error message:
```
s6-mkdir: warning: unable to mkdir /var/run/s6: Permission denied
```
This certainly is possible - please have a look at the[ jupyter/base-notebook](https://github.com/jupyter/docker-stacks/tree/master/base-notebook) which is run with group 0 (safe) to be able to add the arbitrary user's entry to /etc/passwd at runtime - this magic happens here: [start.sh](https://github.com/jupyter/docker-stacks/blob/master/base-notebook/start.sh)

**Comment by cboettig** [comment]
@mirekphd Yes, it certainly is possible.  Please set the env var `USERID` when running your image, as described in the documentation https://www.rocker-project.org/use/managing_users/

We should probably support doing this with `-u` too, the options supported there have evolved quite a bit since we adopted the `USERID` convention in very early versions of Docker and we'll probably need to keep that for backwards compatibility.  PR's welcome, note that the current user config is handled by https://github.com/rocker-org/rocker-versioned/blob/master/rstudio/userconf.sh

**Comment by mirekphd** [comment]
@cboettig This works, but only in docker. It would be a security hole if one could set an arbitrary user ID by means of an environmental variable, which is why OpenShift will not allow it. We still cannot bypass the requirement of setting user ID properly, using OpenShift's "runAsUser" (which must also come from a pre-defined range over which local node admins and devs have no control). This is executed using docker run -u switch, which raises the above-mentioned error.

So the proposed solution:
```
docker run -it -p 8787:8787 -e USERID=1234 rocker/rstudio
```
... while it sets uid and gid to 1234 under docker, will not be reproducible under OpenShift, because OpenShift always uses the -u switch to control user IDs. So assuming we set runAsUser to 3456 and passed env variable USERID set to 1234, OpenShift will execute:
```
docker run -it -p 8787:8787 -e USERID=1234 -u 3456 rocker/rstudio
```
but an attempt to set -u different from 0 (here: to 3456) will in case of rocker/rstudio result in: "
```
s6-mkdir: warning: unable to mkdir /var/run/s6: Permission denied
```

**Comment by sapkra** [comment]
I had the same problem with another docker container using s6 and found the following repository which already has a fix for it.

https://github.com/untoreh/containers/commit/fd6355b5e34eb0ef29522306b5becc25719e31c4

I hope it will also help you.

**Comment by mirekphd** [comment]
And the /etc/password editing script added to entry point (where the final UID is known) comes verbatim straight from the OpenShift documentation: 
https://docs.openshift.com/container-platform/3.10/creating_images/guidelines.html#openshift-specific-guidelines (see Support Arbitrary User IDs). Make sure you delete the existing rstudio entry first @cboettig

**Comment by cboettig** [comment]
@sapkra @mirekphd thanks, this sounds promising.  I don't have an openshift account to test this though, I'm hesitant to dive in here.  Any interest from folks in sending a PR against https://github.com/rocker-org/rocker-versioned/blob/master/rstudio/ setup to adjust the s6 init process to handle this case?

**Comment by BryanHepworth** [comment]
Hi @cboettig 
I see there have been some changes - I'm looking to get it to run openshift and have enough resource to try this.

**Comment by cboettig** [comment]
Thanks @BryanHepworth , any updates on this? we'd still be happy for a PR that could address the user id issue for OpenShift without breaking existing behavior.

**Comment by BryanHepworth** [comment]
Hi Carl

I’ve had chance to sort out a few things over the Christmas break, hoping to make progress this month.

Bryan

**Comment by priggad** [comment]
Hi, was support for running in Openshift every implemented?

**Comment by cboettig** [comment]
@priggad from the thread it sounds that this is resolved upstream.  if not, can you open a new issue in the currently active repo (https://github.com/rocker-org/rocker-versioned2/issues) describing the problem you see and how to reproduce it, linking back here as appropriate?
